feat: support ocp-indent

zjhmale opened the pull request:
I think it will be super awesome if we can use either official indentation or the ocp-indent :)

Maelan commented:
ocp-indent already ships with its indentation file. _When ocp-indent is available_, and it is properly loaded by the user⁽¹⁾, it takes precedence over the builtin indentation file (because of Vim’s filetype-based loading mechanism, and the variable `b:did_indent`).

I insist on “when it is available”: ocp-indent may be present in the environment and disappear the moment after, for example because you installed it in some opam switch, and then switched to another one. With your patch, if the user enables the option, you override the builtin indentation plugin with ocp-indent even when it does not exist.

Plus, you are duplicating the Vim file from ocp-indent. When ocp-indent is available, you’ll have both files at the same time. This is redundant, and it also means that someone would have to maintain the file here in sync with the file upstream (and what about different versions of ocp-indent?).

In summary, to me, you are creating an unnecessary and undesirable reverse dependency from the builtin files (here) to an external plugin (ocp-indent). It is simply better to let the user load her set of plugins (ocp-indent, but also merlin and what not) herself.

It is true that the builtin indentation file is outdated and flawed (because, well, everyone uses ocp-indent) and something should probably be done about it, but I don’t know what right now (by the way, the builtin folding plugin is even worse and should be nuked, in my opinion).

---

Other than that, `indent/default-indent.vim` is badly named, the name should at least mention OCaml. Remember that, from the point of view of the Vim user, the `indent/` sub-directory of all directories listed in the 'runtimepath' are merged.

---

⁽¹⁾ “Properly loaded” means putting ocp-indent in the 'runtimepath' _before_ builtin files, as in:

```vim
let g:opamshare=substitute(system('opam config var share'),'\n$','','')
" load the ocp-indent plugin, if available:
execute 'set rtp^=' . g:opamshare . '/ocp-indent/vim'
```

zjhmale commented:
Hi @Maelan, thx for your kindly review, I totally accept the points you mentioned without any doubts 😅 

> I insist on “when it is available”: ocp-indent may be present in the environment and disappear the moment after, for example because you installed it in some opam switch, and then switched to another one. With your patch, if the user enables the option, you override the builtin indentation plugin with ocp-indent even when it does not exist.

> Plus, you are duplicating the Vim file from ocp-indent. When ocp-indent is available, you’ll have both files at the same time. This is redundant, and it also means that someone would have to maintain the file here in sync with the file upstream (and what about different versions of ocp-indent?).

That's right, it will be a lot of trouble if a user enables the flag I introduced when there is no ocp-indent actually installed, and maintain an extra ocp-indent.vim file will be tedious to avoid conflicts with the upstream ocp-indent file.

> Other than that, indent/default-indent.vim is badly named, the name should at least mention OCaml. Remember that, from the point of view of the Vim user, the indent/ sub-directory of all directories listed in the 'runtimepath' are merged.

Initially, I named it as indent/ocaml-indent.vim but later then for some reason, I changed to indent/default-indent.vim since I thought the `default` may match the semantic for the indentation behavior when a user use this plugin without any extra flags enabled, but yeah, you are right, from a user's perspective, it will be confusing to find a `default-indent` file among a bunch of indent files which named with its corresponding language name.

I will close this, thnx again :)
